### PR TITLE
Constant time equals method

### PR DESCRIPTION
**Pull request by j-baker**
Makes timing attacks impossible if e.g. you have a `Map<AuthHeader, Some permissions granted>`

### CODE REVIEW

File: auth-tokens/src/main/java/com/palantir/tokens/auth/BearerToken.java [hunk 1]
```diff
@@ -39,6 +41,12 @@
     @JsonValue
     public abstract String getToken();
 
+    @Value.Derived
+    @Value.Auxiliary
+    byte[] getTokenAsBytes() {
```
Comment (iamdanfox) [inline auth-tokens/src/main/java/com/palantir/tokens/auth/BearerToken.java]: Any reason this is package accessible or could this be made `protected`?

Comment (j-baker) [reply auth-tokens/src/main/java/com/palantir/tokens/auth/BearerToken.java]: protected is actually more accessible than package?

Comment (j-baker) [reply auth-tokens/src/main/java/com/palantir/tokens/auth/BearerToken.java]: protected = package but also subclasses can access

**Comment by iamdanfox** [comment]
Just a thought - this is going to double the memory footprint of BearerTokens right?

File: auth-tokens/src/main/java/com/palantir/tokens/auth/BearerToken.java [hunk 2]
```diff
@@ -39,6 +41,12 @@
     @JsonValue
     public abstract String getToken();
 
+    @Value.Derived
+    @Value.Auxiliary
```
Comment (iamdanfox) [inline auth-tokens/src/main/java/com/palantir/tokens/auth/BearerToken.java]: Given that we already have our own `equals` `hashCode` and `toString` - this `@Value.Auxiliary` doesn't actually do anything right? https://immutables.github.io/immutable.html#auxiliary-attributes

**Comment by j-baker** [comment]
yes, but I think the extra 200 bytes ram is worth it. I could also do the conversion on equals, but this is an expensive conversion according to profiling i've done in the past. To be clear, right now it's a bit risky to have the cache I describe - on services with little traffic it's likely possible you can infer bearer tokens remotely right now.

**Review by iamdanfox — APPROVED** [review]

**Comment by iamdanfox** [comment]
Looks like I don't have merge perms on this repo - maybe @uschi2000 can help?

**Comment by j-baker** [comment]
Good news! I do.

**Pull request merged**